File: protocol/python/tool_registry_manager.py
"""
Tool Registry Manager for Cross-Language Tool Execution
Manages tool registration, discovery, and metadata for both Python and TypeScript tools
"""

import asyncio
import json
from typing import Dict, List, Optional, Any, Set, Callable
from dataclasses import dataclass, asdict
from datetime import datetime
from pathlib import Path
import logging

from .types import (
    ToolMetadata,
    ToolCapability,
    ToolExecutionRequest,
    ToolExecutionResult,
    Language,
    PROTOCOL_VERSION
)
from .registry import ToolRegistry
from .bridge import ExecutionBridge

logger = logging.getLogger(__name__)

@dataclass
class RegisteredTool:
    """Represents a registered tool with full metadata"""
    id: str
    name: str
    description: str
    language: Language
    category: str
    version: str
    input_schema: Dict[str, Any]
    output_schema: Optional[Dict[str, Any]]
    capabilities: List[ToolCapability]
    dependencies: List[str]
    examples: List[Dict[str, Any]]
    handler: Optional[Callable] = None
    remote: bool = False
    registered_at: datetime = None
    last_used: Optional[datetime] = None
    execution_count: int = 0
    average_execution_time: float = 0.0
    
    def __post_init__(self):
        if self.registered_at is None:
            self.registered_at = datetime.now()
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for serialization"""
        data = asdict(self)
        data['registered_at'] = self.registered_at.isoformat()
        if self.last_used:
            data['last_used'] = self.last_used.isoformat()
        data.pop('handler', None)  # Don't serialize handler
        return data
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'RegisteredTool':
        """Create from dictionary"""
        data = data.copy()
        data['registered_at'] = datetime.fromisoformat(data['registered_at'])
        if data.get('last_used'):
            data['last_used'] = datetime.fromisoformat(data['last_used'])
        return cls(**data)

class ToolRegistryManager:
    """
    Central registry manager for cross-language tool execution
    Manages tool registration, discovery, and metadata synchronization
    """
    
    def __init__(self, language: Language = Language.PYTHON):
        self.language = language
        self.tools: Dict[str, RegisteredTool] = {}
        self.remote_tools: Dict[str, RegisteredTool] = {}
        self.capabilities: Dict[str, Set[str]] = {}  # capability -> tool_ids
        self.categories: Dict[str, Set[str]] = {}    # category -> tool_ids
        self.bridge: Optional[ExecutionBridge] = None
        self._sync_lock = asyncio.Lock()
        self._initialized = False
        
    async def initialize(self, bridge: Optional[ExecutionBridge] = None) -> None:
        """Initialize the registry manager"""
        if self._initialized:
            return
            
        self.bridge = bridge or ExecutionBridge()
        
        # Load local tools
        await self._load_local_tools()
        
        # Sync with remote registry if bridge is available
        if self.bridge:
            await self._sync_remote_tools()
            
        self._initialized = True
        logger.info(f"ToolRegistryManager initialized with {len(self.tools)} local tools and {len(self.remote_tools)} remote tools")
        
    async def _load_local_tools(self) -> None:
        """Load tools from the local registry"""
        # Import local tool modules dynamically
        if self.language == Language.PYTHON:
            await self._load_python_tools()
        else:
            await self._load_typescript_tools()
            
    async def _load_python_tools(self) -> None:
        """Load Python tools from dgm/tools directory"""
        tools_dir = Path(__file__).parent.parent.parent / "dgm" / "tools"
        if not tools_dir.exists():
            logger.warning(f"Tools directory not found: {tools_dir}")
            return
            
        for tool_file in tools_dir.glob("*.py"):
            if tool_file.name == "__init__.py":
                continue
                
            try:
                # Dynamically import tool module
                module_name = tool_file.stem
                module = __import__(f"dgm.tools.{module_name}", fromlist=["tool_info", "tool_function"])
                
                if hasattr(module, "tool_info"):
                    tool_info = module.tool_info()
                    tool_id = tool_info.get("name", module_name)
                    
                    # Create RegisteredTool
                    tool = RegisteredTool(
                        id=tool_id,
                        name=tool_info.get("name", tool_id),
                        description=tool_info.get("description", ""),
                        language=Language.PYTHON,
                        category=tool_info.get("category", "utility"),
                        version=tool_info.get("version", "1.0.0"),
                        input_schema=tool_info.get("input_schema", {}),
                        output_schema=tool_info.get("output_schema"),
                        capabilities=tool_info.get("capabilities", []),
                        dependencies=tool_info.get("dependencies", []),
                        examples=tool_info.get("examples", []),
                        handler=getattr(module, "tool_function", None),
                        remote=False
                    )
                    
                    await self.register_tool(tool)
                    
            except Exception as e:
                logger.error(f"Failed to load tool from {tool_file}: {e}")
                
    async def _load_typescript_tools(self) -> None:
        """Load TypeScript tools via bridge"""
        # TypeScript tools are loaded via the bridge
        pass
        
    async def register_tool(self, tool: RegisteredTool) -> None:
        """Register a tool in the registry"""
        async with self._sync_lock:
            self.tools[tool.id] = tool
            
            # Update capability index
            for capability in tool.capabilities:
                if capability not in self.capabilities:
                    self.capabilities[capability] = set()
                self.capabilities[capability].add(tool.id)
                
            # Update category index
            if tool.category not in self.categories:
                self.categories[tool.category] = set()
            self.categories[tool.category].add(tool.id)
            
            logger.info(f"Registered tool: {tool.id} ({tool.language})")
            
    async def unregister_tool(self, tool_id: str) -> None:
        """Unregister a tool from the registry"""
        async with self._sync_lock:
            if tool_id in self.tools:
                tool = self.tools[tool_id]
                
                # Remove from capability index
                for capability in tool.capabilities:
                    if capability in self.capabilities:
                        self.capabilities[capability].discard(tool_id)
                        
                # Remove from category index
                if tool.category in self.categories:
                    self.categories[tool.category].discard(tool_id)
                    
                del self.tools[tool_id]
                logger.info(f"Unregistered tool: {tool_id}")
                
    async def get_tool(self, tool_id: str) -> Optional[RegisteredTool]:
        """Get a tool by ID"""
        # Check local tools first
        if tool_id in self.tools:
            return self.tools[tool_id]
            
        # Check remote tools
        if tool_id in self.remote_tools:
            return self.remote_tools[tool_id]
            
        return None
        
    async def list_tools(
        self,
        category: Optional[str] = None,
        language: Optional[Language] = None,
        capabilities: Optional[List[str]] = None
    ) -> List[RegisteredTool]:
        """List tools with optional filters"""
        tools = list(self.tools.values()) + list(self.remote_tools.values())
        
        # Filter by category
        if category:
            tools = [t for t in tools if t.category == category]
            
        # Filter by language
        if language:
            tools = [t for t in tools if t.language == language]
            
        # Filter by capabilities
        if capabilities:
            required_caps = set(capabilities)
            tools = [t for t in tools if required_caps.issubset(set(t.capabilities))]
            
        return tools
        
    async def search_tools(self, query: str) -> List[RegisteredTool]:
        """Search tools by name or description"""
        query_lower = query.lower()
        results = []
        
        for tool in list(self.tools.values()) + list(self.remote_tools.values()):
            if (query_lower in tool.name.lower() or 
                query_lower in tool.description.lower() or
                any(query_lower in cap.lower() for cap in tool.capabilities)):
                results.append(tool)
                
        return results
        
    async def get_tools_by_capability(self, capability: str) -> List[RegisteredTool]:
        """Get all tools with a specific capability"""
        tool_ids = self.capabilities.get(capability, set())
        tools = []
        
        for tool_id in tool_ids:
            tool = await self.get_tool(tool_id)
            if tool:
                tools.append(tool)
                
        return tools
        
    async def _sync_remote_tools(self) -> None:
        """Synchronize with remote tool registry"""
        if not self.bridge:
            return
            
        try:
            # Request tool list from remote registry
            request = {
                "jsonrpc": "2.0",
                "method": "registry.list_tools",
                "params": {
                    "language": self.language.value
                },
                "id": "sync-" + str(datetime.now().timestamp())
            }
            
            response = await self.bridge.send_request(request)
            
            if "result" in response:
                remote_tools = response["result"]["tools"]
                
                # Update remote tools registry
                self.remote_tools.clear()
                for tool_data in remote_tools:
                    tool = RegisteredTool.from_dict(tool_data)
                    tool.remote = True
                    self.remote_tools[tool.id] = tool
                    
                    # Update indices
                    for capability in tool.capabilities:
                        if capability not in self.capabilities:
                            self.capabilities[capability] = set()
                        self.capabilities[capability].add(tool.id)
                        
                    if tool.category not in self.categories:
                        self.categories[tool.category] = set()
                    self.categories[tool.category].add(tool.id)
                    
                logger.info(f"Synchronized {len(self.remote_tools)} remote tools")
                
        except Exception as e:
            logger.error(f"Failed to sync remote tools: {e}")
            
    async def execute_tool(
        self,
        tool_id: str,
        parameters: Any,
        context: Optional[Dict[str, Any]] = None
    ) -> ToolExecutionResult:
        """Execute a tool by ID"""
        tool = await self.get_tool(tool_id)
        if not tool:
            raise ValueError(f"Tool not found: {tool_id}")
            
        # Update usage statistics
        tool.last_used = datetime.now()
        tool.execution_count += 1
        start_time = asyncio.get_event_loop().time()
        
        try:
            if tool.remote:
                # Execute via bridge
                result = await self.bridge.execute(
                    tool_id,
                    tool.language,
                    parameters,
                    context
                )
            else:
                # Execute locally
                if not tool.handler:
                    raise ValueError(f"Tool {tool_id} has no handler")
                    
                result = await tool.handler(parameters, context)
                
            # Update execution time statistics
            execution_time = asyncio.get_event_loop().time() - start_time
            tool.average_execution_time = (
                (tool.average_execution_time * (tool.execution_count - 1) + execution_time) /
                tool.execution_count
            )
            
            return result
            
        except Exception as e:
            logger.error(f"Tool execution failed for {tool_id}: {e}")
            raise
            
    async def get_tool_statistics(self, tool_id: str) -> Dict[str, Any]:
        """Get execution statistics for a tool"""
        tool = await self.get_tool(tool_id)
        if not tool:
            return {}
            
        return {
            "id": tool.id,
            "execution_count": tool.execution_count,
            "average_execution_time": tool.average_execution_time,
            "last_used": tool.last_used.isoformat() if tool.last_used else None,
            "registered_at": tool.registered_at.isoformat()
        }
        
    async def export_registry(self, file_path: str) -> None:
        """Export registry to a JSON file"""
        registry_data = {
            "version": PROTOCOL_VERSION,
            "language": self.language.value,
            "tools": [tool.to_dict() for tool in self.tools.values()],
            "remote_tools": [tool.to_dict() for tool in self.remote_tools.values()],
            "exported_at": datetime.now().isoformat()
        }
        
        with open(file_path, 'w') as f:
            json.dump(registry_data, f, indent=2)
            
        logger.info(f"Exported registry to {file_path}")
        
    async def import_registry(self, file_path: str) -> None:
        """Import registry from a JSON file"""
        with open(file_path, 'r') as f:
            registry_data = json.load(f)
            
        # Clear existing registry
        self.tools.clear()
        self.remote_tools.clear()
        self.capabilities.clear()
        self.categories.clear()
        
        # Import tools
        for tool_data in registry_data.get("tools", []):
            tool = RegisteredTool.from_dict(tool_data)
            await self.register_tool(tool)
            
        for tool_data in registry_data.get("remote_tools", []):
            tool = RegisteredTool.from_dict(tool_data)
            tool.remote = True
            self.remote_tools[tool.id] = tool
            
        logger.info(f"Imported registry from {file_path}")

# Singleton instance
_registry_manager: Optional[ToolRegistryManager] = None

def get_registry_manager() -> ToolRegistryManager:
    """Get the singleton registry manager instance"""
    global _registry_manager
    if _registry_manager is None:
        _registry_manager = ToolRegistryManager()
    return _registry_manager

File: protocol/typescript/src/tool-registry-manager.ts
/**
 * Tool Registry Manager for Cross-Language Tool Execution
 * Manages tool registration, discovery, and metadata for both TypeScript and Python tools
 */

import { 
  ToolMetadata,
  ToolCapability,
  ToolExecutionRequest,
  ToolExecutionResult,
  Language,
  PROTOCOL_VERSION
} from './types'
import { ToolRegistry } from './registry'
import { ExecutionBridge } from './bridge'
import * as fs from 'fs/promises'
import * as path from 'path'

export interface RegisteredTool {
  id: string
  name: string
  description: string
  language: Language
  category: string
  version: string
  inputSchema: Record<string, any>
  outputSchema?: Record<string, any>
  capabilities: ToolCapability[]
  dependencies: string[]
  examples: Array<Record<string, any>>
  handler?: (params: any, context: any) => Promise<any>
  remote: boolean
  registeredAt: Date
  lastUsed?: Date
  executionCount: number
  averageExecutionTime: number
}

export interface ToolFilter {
  category?: string
  language?: Language
  capabilities?: string[]
}

export interface ToolStatistics {
  id: string
  executionCount: number
  averageExecutionTime: number
  lastUsed?: string
  registeredAt: string
}

export class ToolRegistryManager {
  private static instance: ToolRegistryManager | null = null
  
  private language: Language
  private tools: Map<string, RegisteredTool> = new Map()
  private remoteTools: Map<string, RegisteredTool> = new Map()
  private capabilities: Map<string, Set<string>> = new Map() // capability -> tool_ids
  private categories: Map<string, Set<string>> = new Map()   // category -> tool_ids
  private bridge?: ExecutionBridge
  private syncLock = false
  private initialized = false
  
  private constructor(language: Language = Language.TYPESCRIPT) {
    this.language = language
  }
  
  /**
   * Get singleton instance
   */
  static getInstance(language?: Language): ToolRegistryManager {
    if (!ToolRegistryManager.instance) {
      ToolRegistryManager.instance = new ToolRegistryManager(language)
    }
    return ToolRegistryManager.instance
  }
  
  /**
   * Initialize the registry manager
   */
  async initialize(bridge?: ExecutionBridge): Promise<void> {
    if (this.initialized) {
      return
    }
    
    this.bridge = bridge || new ExecutionBridge()
    
    // Load local tools
    await this.loadLocalTools()
    
    // Sync with remote registry if bridge is available
    if (this.bridge) {
      await this.syncRemoteTools()
    }
    
    this.initialized = true
    console.log(`ToolRegistryManager initialized with ${this.tools.size} local tools and ${this.remoteTools.size} remote tools`)
  }
  
  /**
   * Load tools from the local registry
   */
  private async loadLocalTools(): Promise<void> {
    if (this.language === Language.TYPESCRIPT) {
      await this.loadTypeScriptTools()
    } else {
      await this.loadPythonTools()
    }
  }
  
  /**
   * Load TypeScript tools from opencode/packages/opencode/src/tool directory
   */
  private async loadTypeScriptTools(): Promise<void> {
    const toolsDir = path.join(__dirname, '../../../../opencode/packages/opencode/src/tool')
    
    try {
      const files = await fs.readdir(toolsDir)
      
      for (const file of files) {
        if (!file.endsWith('.ts') || file === 'tool.ts') {
          continue
        }
        
        try {
          // Dynamically import tool module
          const modulePath = path.join(toolsDir, file)
          const module = await import(modulePath)
          
          // Look for tool exports
          for (const exportName of Object.keys(module)) {
            const exported = module[exportName]
            
            if (exported && typeof exported === 'object' && exported.id) {
              // Create RegisteredTool
              const tool: RegisteredTool = {
                id: exported.id,
                name: exported.name || exported.id,
                description: exported.description || '',
                language: Language.TYPESCRIPT,
                category: exported.category || 'utility',
                version: exported.version || '1.0.0',
                inputSchema: exported.parameters || {},
                outputSchema: exported.outputSchema,
                capabilities: exported.capabilities || [],
                dependencies: exported.dependencies || [],
                examples: exported.examples || [],
                handler: exported.execute,
                remote: false,
                registeredAt: new Date(),
                executionCount: 0,
                averageExecutionTime: 0
              }
              
              await this.registerTool(tool)
            }
          }
        } catch (error) {
          console.error(`Failed to load tool from ${file}:`, error)
        }
      }
    } catch (error) {
      console.warn(`Tools directory not found: ${toolsDir}`)
    }
  }
  
  /**
   * Load Python tools via bridge
   */
  private async loadPythonTools(): Promise<void> {
    // Python tools are loaded via the bridge
  }
  
  /**
   * Register a tool in the registry
   */
  async registerTool(tool: RegisteredTool): Promise<void> {
    while (this.syncLock) {
      await new Promise(resolve => setTimeout(resolve, 10))
    }
    
    this.syncLock = true
    try {
      this.tools.set(tool.id, tool)
      
      // Update capability index
      for (const capability of tool.capabilities) {
        if (!this.capabilities.has(capability)) {
          this.capabilities.set(capability, new Set())
        }
        this.capabilities.get(capability)!.add(tool.id)
      }
      
      // Update category index
      if (!this.categories.has(tool.category)) {
        this.categories.set(tool.category, new Set())
      }
      this.categories.get(tool.category)!.add(tool.id)
      
      console.log(`Registered tool: ${tool.id} (${tool.language})`)
    } finally {
      this.syncLock = false
    }
  }
  
  /**
   * Unregister a tool from the registry
   */
  async unregisterTool(toolId: string): Promise<void> {
    while (this.syncLock) {
      await new Promise(resolve => setTimeout(resolve, 10))
    }
    
    this.syncLock = true
    try {
      const tool = this.tools.get(toolId)
      if (tool) {
        // Remove from capability index
        for (const capability of tool.capabilities) {
          this.capabilities.get(capability)?.delete(toolId)
        }
        
        // Remove from category index
        this.categories.get(tool.category)?.delete(toolId)
        
        this.tools.delete(toolId)
        console.log(`Unregistered tool: ${toolId}`)
      }
    } finally {
      this.syncLock = false
    }
  }
  
  /**
   * Get a tool by ID
   */
  async getTool(toolId: string): Promise<RegisteredTool | undefined> {
    // Check local tools first
    if (this.tools.has(toolId)) {
      return this.tools.get(toolId)
    }
    
    // Check remote tools
    if (this.remoteTools.has(toolId)) {
      return this.remoteTools.get(toolId)
    }
    
    return undefined
  }
  
  /**
   * List tools with optional filters
   */
  async listTools(filter?: ToolFilter): Promise<RegisteredTool[]> {
    let tools = [...this.tools.values(), ...this.remoteTools.values()]
    
    // Filter by category
    if (filter?.category) {
      tools = tools.filter(t => t.category === filter.category)
    }
    
    // Filter by language
    if (filter?.language) {
      tools = tools.filter(t => t.language === filter.language)
    }
    
    // Filter by capabilities
    if (filter?.capabilities && filter.capabilities.length > 0) {
      const requiredCaps = new Set(filter.capabilities)
      tools = tools.filter(t => {
        const toolCaps = new Set(t.capabilities)
        for (const cap of requiredCaps) {
          if (!toolCaps.has(cap as ToolCapability)) {
            return false
          }
        }
        return true
      })
    }
    
    return tools
  }
  
  /**
   * Search tools by name or description
   */
  async searchTools(query: string): Promise<RegisteredTool[]> {
    const queryLower = query.toLowerCase()
    const results: RegisteredTool[] = []
    
    for (const tool of [...this.tools.values(), ...this.remoteTools.values()]) {
      if (tool.name.toLowerCase().includes(queryLower) ||
          tool.description.toLowerCase().includes(queryLower) ||
          tool.capabilities.some(cap => cap.toLowerCase().includes(queryLower))) {
        results.push(tool)
      }
    }
    
    return results
  }
  
  /**
   * Get all tools with a specific capability
   */
  async getToolsByCapability(capability: string): Promise<RegisteredTool[]> {
    const toolIds = this.capabilities.get(capability) || new Set()
    const tools: RegisteredTool[] = []
    
    for (const toolId of toolIds) {
      const tool = await this.getTool(toolId)
      if (tool) {
        tools.push(tool)
      }
    }
    
    return tools
  }
  
  /**
   * Synchronize with remote tool registry
   */
  private async syncRemoteTools(): Promise<void> {
    if (!this.bridge) {
      return
    }
    
    try {
      // Request tool list from remote registry
      const request = {
        jsonrpc: '2.0',
        method: 'registry.list_tools',
        params: {
          language: this.language
        },
        id: `sync-${Date.now()}`
      }
      
      const response = await this.bridge.sendRequest(request)
      
      if (response.result) {
        const remoteTools = response.result.tools as any[]
        
        // Update remote tools registry
        this.remoteTools.clear()
        for (const toolData of remoteTools) {
          const tool: RegisteredTool = {
            ...toolData,
            remote: true,
            registeredAt: new Date(toolData.registeredAt),
            lastUsed: toolData.lastUsed ? new Date(toolData.lastUsed) : undefined
          }
          
          this.remoteTools.set(tool.id, tool)
          
          // Update indices
          for (const capability of tool.capabilities) {
            if (!this.capabilities.has(capability)) {
              this.capabilities.set(capability, new Set())
            }
            this.capabilities.get(capability)!.add(tool.id)
          }
          
          if (!this.categories.has(tool.category)) {
            this.categories.set(tool.category, new Set())
          }
          this.categories.get(tool.category)!.add(tool.id)
        }
        
        console.log(`Synchronized ${this.remoteTools.size} remote tools`)
      }
    } catch (error) {
      console.error('Failed to sync remote tools:', error)
    }
  }
  
  /**
   * Execute a tool by ID
   */
  async executeTool(
    toolId: string,
    parameters: any,
    context?: Record<string, any>
  ): Promise<ToolExecutionResult> {
    const tool = await this.getTool(toolId)
    if (!tool) {
      throw new Error(`Tool not found: ${toolId}`)
    }
    
    // Update usage statistics
    tool.lastUsed = new Date()
    tool.executionCount++
    const startTime = Date.now()
    
    try {
      let result: any
      
      if (tool.remote) {
        // Execute via bridge
        result = await this.bridge!.execute(
          toolId,
          tool.language,
          parameters,
          context || {}
        )
      } else {
        // Execute locally
        if (!tool.handler) {
          throw new Error(`Tool ${toolId} has no handler`)
        }
        
        result = await tool.handler(parameters, context)
      }
      
      // Update execution time statistics
      const executionTime = Date.now() - startTime
      tool.averageExecutionTime = 
        (tool.averageExecutionTime * (tool.executionCount - 1) + executionTime) /
        tool.executionCount
      
      return result
    } catch (error) {
      console.error(`Tool execution failed for ${toolId}:`, error)
      throw error
    }
  }
  
  /**
   * Get execution statistics for a tool
   */
  async getToolStatistics(toolId: string): Promise<ToolStatistics | null> {
    const tool = await this.getTool(toolId)
    if (!tool) {
      return null
    }
    
    return {
      id: tool.id,
      executionCount: tool.executionCount,
      averageExecutionTime: tool.averageExecutionTime,
      lastUsed: tool.lastUsed?.toISOString(),
      registeredAt: tool.registeredAt.toISOString()
    }
  }
  
  /**
   * Export registry to a JSON file
   */
  async exportRegistry(filePath: string): Promise<void> {
    const registryData = {
      version: PROTOCOL_VERSION,
      language: this.language,
      tools: Array.from(this.tools.values()).map(tool => this.toolToJson(tool)),
      remoteTools: Array.from(this.remoteTools.values()).map(tool => this.toolToJson(tool)),
      exportedAt: new Date().toISOString()
    }
    
    await fs.writeFile(filePath, JSON.stringify(registryData, null, 2))
    console.log(`Exported registry to ${filePath}`)
  }
  
  /**
   * Import registry from a JSON file
   */
  async importRegistry(filePath: string): Promise<void> {
    const data = await fs.readFile(filePath, 'utf-8')
    const registryData = JSON.parse(data)
    
    // Clear existing registry
    this.tools.clear()
    this.remoteTools.clear()
    this.capabilities.clear()
    this.categories.clear()
    
    // Import tools
    for (const toolData of registryData.tools || []) {
      const tool = this.toolFromJson(toolData)
      await this.registerTool(tool)
    }
    
    for (const toolData of registryData.remoteTools || []) {
      const tool = this.toolFromJson(toolData)
      tool.remote = true
      this.remoteTools.set(tool.id, tool)
    }
    
    console.log(`Imported registry from ${filePath}`)
  }
  
  /**
   * Convert tool to JSON-serializable format
   */
  private toolToJson(tool: RegisteredTool): any {
    const { handler, ...data } = tool
    return {
      ...data,
      registeredAt: tool.registeredAt.toISOString(),
      lastUsed: tool.lastUsed?.toISOString()
    }
  }
  
  /**
   * Convert JSON data to tool
   */
  private toolFromJson(data: any): RegisteredTool {
    return {
      ...data,
      registeredAt: new Date(data.registeredAt),
      lastUsed: data.lastUsed ? new Date(data.lastUsed) : undefined
    }
  }
}